### PR TITLE
Explain page: add timeline visualization

### PR DESCRIPTION
**Pull request by yoomlam**
Innovation project: [Appeal Narrative](https://dsva.slack.com/archives/C022T6KSE6S/p1622571159006000)

### Description
On the `explain` page, added timeline visualization to accompany narrative table.

### Acceptance Criteria
- [ ] Code compiles correctly
- [ ] `explain` page shows timeline visualization for appeal

### Testing Plan
Install new javascript library: `cd client; yarn install`

Insert `binding.pry` into lines like 113 or 142 in `spec/feature/explain_spec.rb` and run the rspec. 
Should see something like this on the `explain` page:
image

Use your mouse to zoom in/out (pinch or scrollwheel) and pan (drag) along the timeline.

### CODE REVIEW

**Comment by yoomlam** [comment]
With `binding.pry` at spec/feature/explain_spec.rb:142

image

File: app/assets/javascripts/explain-appeal.js [hunk 1]
```diff
@@ -1,6 +1,109 @@
...
+    className: (item)=> item.className + " hearing_" + item.status
+  }
+};
+
+// https://visjs.github.io/vis-timeline/docs/timeline/#items
+function decorateTimelineItems(items){
+  items.forEach(item => {
+    item.className = item.record_type
+
+    if(!itemDecoration.hasOwnProperty(item.table_name)) return;
```
Comment (yoomlam) [inline app/assets/javascripts/explain-appeal.js:63]: `item.table_name` returns `tasks` for task events. `tasks` is a property on the `itemDecoration` constant.

File: app/assets/javascripts/explain-appeal.js [hunk 2]
```diff
@@ -1,6 +1,109 @@
...
+function addTimeline(elementId, timeline_data){
+  const timelineElement = document.getElementById(elementId);
+  // const items = new vis.DataSet(decorateTimelineItems(timeline_data));
+  items.add(decorateTimelineItems(timeline_data))
+  const groups = groupEventItems(timeline_data);
+  const startDates = items.get({fields: ['start']}).map((str)=> new Date(str.start));
+  const endDates = items.get({fields: ['end']}).map((str)=> new Date(str.end));
+  const millisInAMonth = 1000*60*60*24*30 // an approximate month in milliseconds
+  const timeline_options = {
+    width: '95%',
```
Comment (yoomlam) [inline app/assets/javascripts/explain-appeal.js:23]: Leave some whitespace on the right to allow user to scroll the page. Scrolling in the timeline area causes the timeline to zoom.

File: app/assets/javascripts/explain-appeal.js [hunk 3]
```diff
@@ -1,6 +1,109 @@
...
+    item.className = item.record_type
+
+    if(!itemDecoration.hasOwnProperty(item.table_name)) return;
+
+    for ([key, value] of Object.entries(itemDecoration[item.table_name])) {
+      if (typeof value === 'function') {
+        value = value(item)
+      }
+      item[key] = value
+    }
```
Comment (yoomlam) [inline app/assets/javascripts/explain-appeal.js:70]: For `item.table_name = tasks`, `key = className ` and `value` is a function that sets the CSS classname based on the task's status.

File: app/controllers/concerns/explain_appeal_events_concern.rb [hunk 4]
```diff
@@ -79,8 +84,8 @@ def request_issues_as_event_data
   # rubocop:enable Metrics/AbcSize
 
   def hearings_as_event_data
-    hearing_days = exported_records(HearingDay).index_by { |req| req["id"] }
-    virtual_hearings = exported_records(VirtualHearing).index_by { |req| req["hearing_id"] }
+    hearing_days = exported_records(HearingDay).index_by { |rec| rec["id"] }
+    virtual_hearings = exported_records(VirtualHearing).index_by { |rec| rec["hearing_id"] }
```
Comment (yoomlam) [inline app/controllers/concerns/explain_appeal_events_concern.rb:88]: Fix copy-paste lint.

File: app/controllers/concerns/explain_appeal_events_concern.rb [hunk 5]
```diff
@@ -90,10 +95,6 @@ def hearings_as_event_data
 
   private
 
-  def exported_records(klass)
-    sje.records_hash[klass.table_name] || []
-  end
-
```
Comment (yoomlam) [inline app/controllers/concerns/explain_appeal_events_concern.rb:96]: Moved to explain_controller.

File: spec/feature/explain_spec.rb [hunk 6]
```diff
@@ -100,34 +100,62 @@
...
+        scroll_to(page.find("h3", text: "Timeline"))
+        # binding.pry
+        expect(page).to have_content("show_pii = false")
+        expect(page).to have_content("status: dispatched")
+
+        expect(page).to have_content("priority: false (AOD: false, CAVC: false)")
+        expect(page).to have_content("Intake (no PII)")
+        expect(page).to have_content("Hearing (no PII)")
+        expect(page).to have_content("show_pii: false")
+        expect(page).to have_content("Appeal Narrative (showing PII)")
```
Comment (yoomlam) [inline spec/feature/explain_spec.rb:121]: Refactoring to put existing test into a `context` to distinguish from new `context`s.

File: app/assets/javascripts/explain-appeal.js [hunk 7]
```diff
@@ -1,6 +1,115 @@
...
+// https://visjs.github.io/vis-timeline/docs/timeline/#items
+function decorateTimelineItems(items){
+  items.forEach(item => {
+    item.className = item.record_type;
+
+    if(itemDecoration.hasOwnProperty(item.table_name)){
+      for ([key, value] of Object.entries(itemDecoration[item.table_name])) {
+        item[key] = (typeof value === 'function') ? value(item) : value;
+      }
+    }
```
Comment (yoomlam) [inline app/assets/javascripts/explain-appeal.js:73]: `item.table_name` returns `tasks` for task events. `tasks` is a property on the `itemDecoration` constant.
For `item.table_name = "tasks"`, `key = "className"` and `value` is a function that sets the CSS classname based on the task's status.

File: spec/feature/explain_spec.rb [hunk 8]
```diff
@@ -119,34 +119,62 @@
...
-
-      click_link("toggle show_pii")
-      expect(page).to have_content("show_pii = true")
-      expect(page).to have_content("Intake (showing PII)")
-      expect(page).to have_content("Hearing (showing PII)")
-      expect(page).to have_content("show_pii: true")
-      expect(page).to have_content("Appeal Narrative (contains PII)")
-      page.find("#narrative_table").click
-      task = real_appeal.tasks.sample
-      expect(page).to have_content("#{task.type}_#{task.id}")
```
Comment (yoomlam) [inline spec/feature/explain_spec.rb:146]: Refactoring to put existing test into a `context` to distinguish from new `context`s.

File: app/assets/javascripts/explain-appeal.js [hunk 9]
```diff
@@ -1,6 +1,115 @@
...
+    content: 'others',
+    className: 'group_others',
+    order: 9,
+  });
+  return groups;
+}
+
+function toggleTimelineEventGroup(checkbox, group_id){
+  groups.update({id: group_id, visible: checkbox.checked});
+}
```
Comment (yoomlam) [inline app/assets/javascripts/explain-appeal.js:115]: Used by checkboxes to hide groups in timeline.

File: app/views/explain/show.html.erb [hunk 10]
```diff
@@ -40,14 +40,14 @@
...
+    <summary id="treee_config" style="color: blue">Task tree configuration</summary>
 
     <div id="treeefields" style="width: 200px; height: 300px; overflow-y: scroll; padding:10px">
       <% available_fields.each do |field_name| %>
         <% if treee_fields.include?(field_name.to_sym) %>
-          <input type="checkbox" name="<%=field_name%>" id="<%=field_name%>" checked />
+          <input type="checkbox" id="<%=field_name%>" checked />
         <% else %>
-          <input type="checkbox" name="<%=field_name%>" id="<%=field_name%>" />
+          <input type="checkbox" id="<%=field_name%>" />
```
Comment (yoomlam) [inline app/views/explain/show.html.erb:50]: clean up

**Review by lowellrex — APPROVED** [review]
LGTM!

**Pull request merged**